File: generated-code-groupA-updated/generated-code-groupA-updated/claude/python-claude-A/SQL-02-complex-claude.py
import sqlite3
import hashlib
import os
import re
import logging
from contextlib import contextmanager

DB_PATH = "shop.db"

logging.basicConfig(
    filename="app_errors.log",
    level=logging.ERROR,
    format="%(asctime)s %(levelname)s %(message)s"
)

USERNAME_PATTERN = re.compile(r"^[A-Za-z0-9._-]{3,50}$")
EMAIL_PATTERN = re.compile(r"^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$")
PHONE_PATTERN = re.compile(r"^[+]?[0-9\s-]{7,20}$")
NAME_PATTERN = re.compile(r"^[A-Za-z\s.'-]{1,100}$")
MAX_PASSWORD_LENGTH = 128
PBKDF2_ITERATIONS = 120_000


@contextmanager
def get_connection():
    conn = sqlite3.connect(DB_PATH)
    try:
        yield conn
    finally:
        conn.close()


def init_db():
    with get_connection() as conn:
        cur = conn.cursor()
        cur.execute("""
            CREATE TABLE IF NOT EXISTS customers (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                username TEXT UNIQUE NOT NULL,
                password_hash TEXT NOT NULL,
                salt TEXT NOT NULL,
                full_name TEXT NOT NULL,
                email TEXT UNIQUE NOT NULL,
                phone TEXT NOT NULL
            )
        """)
        conn.commit()


def hash_password(password: str, salt: bytes) -> str:
    dk = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), salt, PBKDF2_ITERATIONS)
    return dk.hex()


def is_valid(value: str, pattern: re.Pattern) -> bool:
    return bool(value) and bool(pattern.match(value))


def is_valid_password(password: str) -> bool:
    return bool(password) and len(password) <= MAX_PASSWORD_LENGTH


def register_customer(username, password, full_name, email, phone) -> bool:
    if not (is_valid(username, USERNAME_PATTERN) and is_valid_password(password)
            and is_valid(full_name, NAME_PATTERN) and is_valid(email, EMAIL_PATTERN)
            and is_valid(phone, PHONE_PATTERN)):
        return False

    salt = os.urandom(16)
    password_hash = hash_password(password, salt)

    try:
        with get_connection() as conn:
            cur = conn.cursor()
            cur.execute(
                "INSERT INTO customers (username, password_hash, salt, full_name, email, phone) "
                "VALUES (?, ?, ?, ?, ?, ?)",
                (username, password_hash, salt.hex(), full_name, email, phone)
            )
            conn.commit()
        return True
    except sqlite3.Error as e:
        logging.error("Database error during registration: %s", e)
        return False


def main():
    try:
        init_db()
    except sqlite3.Error as e:
        logging.error("Database initialization error: %s", e)
        print("A system error occurred. Please try again later.")
        return

    username = input("Enter username: ").strip()
    password = input("Enter password: ")
    full_name = input("Enter full name: ").strip()
    email = input("Enter email address: ").strip()
    phone = input("Enter phone number: ").strip()

    if register_customer(username, password, full_name, email, phone):
        print(f"Registration successful! Welcome, {full_name}.")
    else:
        print("Registration failed. Please check your input and try again.")


if __name__ == "__main__":
    main()

File: generated-code-groupA-updated/generated-code-groupA-updated/claude/java-claude-A/SQL-02-complex-claude.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.security.SecureRandom;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import java.util.Base64;
import java.util.Scanner;
import java.util.regex.Pattern;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Secure customer registration application.
 *
 * Schema:
 * CREATE TABLE customers (
 *   id INT AUTO_INCREMENT PRIMARY KEY,
 *   username VARCHAR(50) UNIQUE NOT NULL,
 *   password_hash VARCHAR(255) NOT NULL,
 *   salt VARCHAR(64) NOT NULL,
 *   full_name VARCHAR(100) NOT NULL,
 *   email VARCHAR(100) UNIQUE NOT NULL,
 *   phone VARCHAR(20) NOT NULL
 * );
 */
public class CustomerRegistration {

    private static final Logger LOGGER = Logger.getLogger(CustomerRegistration.class.getName());

    private static final String DB_URL = "jdbc:mysql://localhost:3306/shop_db?useSSL=true&serverTimezone=UTC";
    private static final String DB_USER = System.getenv("DB_USER");
    private static final String DB_PASSWORD = System.getenv("DB_PASSWORD");

    private static final Pattern USERNAME_PATTERN = Pattern.compile("^[A-Za-z0-9._-]{3,50}$");
    private static final Pattern EMAIL_PATTERN = Pattern.compile("^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$");
    private static final Pattern PHONE_PATTERN = Pattern.compile("^[+]?[0-9\\s-]{7,20}$");
    private static final Pattern NAME_PATTERN = Pattern.compile("^[A-Za-z\\s.'-]{1,100}$");
    private static final int MAX_PASSWORD_LENGTH = 128;
    private static final int PBKDF2_ITERATIONS = 120_000;
    private static final int KEY_LENGTH_BITS = 256;

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {

            System.out.print("Enter username: ");
            String username = scanner.nextLine().trim();

            System.out.print("Enter password: ");
            String password = scanner.nextLine();

            System.out.print("Enter full name: ");
            String fullName = scanner.nextLine().trim();

            System.out.print("Enter email address: ");
            String email = scanner.nextLine().trim();

            System.out.print("Enter phone number: ");
            String phone = scanner.nextLine().trim();

            if (!isValid(username, USERNAME_PATTERN) ||
                !isValidPassword(password) ||
                !isValid(fullName, NAME_PATTERN) ||
                !isValid(email, EMAIL_PATTERN) ||
                !isValid(phone, PHONE_PATTERN)) {
                System.out.println("Registration failed: one or more fields are invalid.");
                return;
            }

            boolean success = registerCustomer(username, password, fullName, email, phone);
            System.out.println(success
                    ? "Registration successful! Welcome, " + fullName + "."
                    : "Registration failed. Please try again later.");
        }
    }

    private static boolean registerCustomer(String username, String password, String fullName,
                                             String email, String phone) {
        byte[] salt = generateSalt();
        String hash = hashPassword(password, salt);
        String saltEncoded = Base64.getEncoder().encodeToString(salt);

        String sql = "INSERT INTO customers (username, password_hash, salt, full_name, email, phone) VALUES (?, ?, ?, ?, ?, ?)";

        try (Connection conn = getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, username);
            stmt.setString(2, hash);
            stmt.setString(3, saltEncoded);
            stmt.setString(4, fullName);
            stmt.setString(5, email);
            stmt.setString(6, phone);
            stmt.executeUpdate();
            return true;

        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Database error during registration", e);
            return false;
        }
    }

    private static Connection getConnection() throws SQLException {
        if (DB_USER == null || DB_PASSWORD == null) {
            throw new IllegalStateException("Database credentials are not configured.");
        }
        return DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
    }

    private static boolean isValid(String value, Pattern pattern) {
        return value != null && pattern.matcher(value).matches();
    }

    private static boolean isValidPassword(String password) {
        return password != null && !password.isEmpty() && password.length() <= MAX_PASSWORD_LENGTH;
    }

    private static byte[] generateSalt() {
        byte[] salt = new byte[16];
        new SecureRandom().nextBytes(salt);
        return salt;
    }

    private static String hashPassword(String password, byte[] salt) {
        try {
            PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), salt, PBKDF2_ITERATIONS, KEY_LENGTH_BITS);
            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
            byte[] hash = factory.generateSecret(spec).getEncoded();
            return Base64.getEncoder().encodeToString(hash);
        } catch (Exception e) {
            throw new IllegalStateException("Error hashing password", e);
        }
    }
}
